README.md

```diff
@@ -263,81 +280,110 @@ collection = ParentEventsTreeCollectionProvider5(events, data_source=data_source
+supports aggregate operations.
 
 ### Terms
 
-- **Data object**: An object of `Data` class which is wrapper under stream. 
+- **Data object**: An instance of `Data` class which is wrapper under stream.
+- **Sequence of elements**:
+  A _Data object_ provides an interface to a sequenced set of values of a specific element type. Stream inside the _Data
+  object_ **don’t actually store** elements; they are computed on demand.
+- **DataSource**:
+  Any source of data. E.g. [Report Data Provider](https://github.com/th2-net/th2-rpt-data-provider), collections,
+  arrays, or I/O resources.
+- **ProviderDataSource**:
+  DataSource object the source of which is [Report Data Provider](https://github.com/th2-net/th2-rpt-data-provider).
+- **SourceAPI**:
+  Each source has its own API to retrieve data. That term/class is used for reference/describe this API.
+- **Commands**:
+  Objects that provide user-friendly interfaces for getting some data from DataSource. Commands use _SourceAPI_ to
+  achieve it.
+- **Adapters**:
+  It's similar to function for `Data.map` method. Adoptable commands used it to update the data stream.
+- **Aggregate operations**:
+  Common operations such as filter, map, limit and so on.
+- **Workflow**: An ordered set of _Aggregate operations_.
+
+### Concept
+
+The library describes the high-level interfaces `ISourceAPI`, `IDataSource`, `ICommand`, `IAdapter`.
+
+Any data source must be described by the `IDataSource` abstract class. These can be _FileDataSource_, _CSVDataSource_, _
+DBDataSource_ and other.
+
+Usually, data sources have some kind of API. Databases - provide SQL language, when working with a file, you can read
+line by line, etc. This API is described by the `ISourceAPI` class. Because different versions of the same data source
+may have different API, it is better to create a class for each version.
+
+Generally, data source APIs are hidden behind convenient interfaces. The role of these interfaces is played
+by `ICommand` classes.
+
+`IAdapter` classes transform data stream like functions for `Data.map` method. Essentially it's the same thing but more
+flexible.
+
+Thus, the native `ProviderDataSource` and the set of commands for it are described. This approach provides great
+opportunities for extension. You can easily create your own unique commands for _ProviderDataSource_, as well as entire
+_DataSource_ classes.
+
+![Data stream pipeline](documentation/img/concept.png)
 
+### Stream operations
 
-- **Sequence of elements**:
-  A _Data object_ provides an interface to a sequenced set of values of a specific element type. 
-  Stream inside the _Data object_ **don’t actually store** elements; they are computed on demand.
+Furthermore, stream operations have two fundamental characteristics that make them very different from collection
+operations: _Pipelining_ and _Internal iteration_.
 
+#### Pipelining
```

Comment:
To be honest, I didn't understand the difference between internal iteration and pipelining)